### PR TITLE
Allow missing fields in structs in ORC

### PR DESCRIPTION
**Pull request by shreyasjoshis**
### Summary

Structs can have missing fields if the the table schema has changed to add nested fields, but the underlying ORC files are not updated. Earlier Presto would fail with an `ArrayIndexOutOfBounds` exception. Now, the missing fields will be marked as null.

This behavior will allow Presto to support addition of fields to structs without needing to change the underlying data.

### Steps to reproduce the problem
When the table schema has struct fields that the underlying ORC file does not have, Presto errors out and fails to read the ORC file. Here are steps to reproduce the problem:

```shell
# Clean table location
bin/hdfs dfs -rm -r -f /user/hive/warehouse/test_schema.db/test_tbl/
bin/hdfs dfs -rm -r -f /user/hive/warehouse/test_schema.db/test_tbl2/
```

Create tables and insert data
```sql
DROP TABLE test_schema.test_tbl;

CREATE EXTERNAL TABLE `test_schema.test_tbl`(
  `a` string,  
  `b` struct<ts:STRING, topic:STRING>)
ROW FORMAT SERDE
  'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
STORED AS INPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
OUTPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat';

INSERT INTO table test_schema.test_tbl SELECT 'a_val1', named_struct('ts', 'b1ts','topic', 'b1topic');

DROP TABLE test_schema.test_tbl2;
CREATE EXTERNAL TABLE `test_schema.test_tbl2`( 
  `a` string,
  `b` struct<ts:STRING>)
ROW FORMAT SERDE
  'org.apache.hadoop.hive.ql.io.orc.OrcSerde'
STORED AS INPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcInputFormat'
OUTPUTFORMAT
  'org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat';

INSERT INTO table test_schema.test_tbl2 SELECT 'a_val2', named_struct('ts', 'b2ts');
```

Copy data over
```
bin/hdfs dfs -cp /user/hive/warehouse/test_schema.db/test_tbl2/000000_0 /user/hive/warehouse/test_schema.db/test_tbl/from_test_tbl2
```

Query from Hive (this works as expected):
```
hive> select * from test_schema.test_tbl;
a_val1	{"ts":"b1ts","topic":"b1topic"}
a_val2	{"ts":"b2ts","topic":null}
```

Query from Presto (This fails with the following):
```java
java.lang.ArrayIndexOutOfBoundsException: 1
	at com.facebook.presto.orc.reader.StructStreamReader.readBlock(StructStreamReader.java:100)
	at com.facebook.presto.orc.OrcRecordReader.readBlock(OrcRecordReader.java:370)
	at com.facebook.presto.hive.orc.OrcPageSource$OrcBlockLoader.load(OrcPageSource.java:240)
	at com.facebook.presto.hive.orc.OrcPageSource$OrcBlockLoader.load(OrcPageSource.java:216)
	at com.facebook.presto.spi.block.LazyBlock.assureLoaded(LazyBlock.java:256)
	at com.facebook.presto.spi.Page.assureLoaded(Page.java:230)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:265)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:337)
	at com.facebook.presto.operator.Driver.lambda$processFor$6(Driver.java:241)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:614)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:235)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:622)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at com.facebook.presto.execution.executor.LegacyPrioritizedSplitRunner.process(LegacyPrioritizedSplitRunner.java:23)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:485)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)```
```

### The solution
With the changes in this PR Presto will output:
```
presto> SELECT * FROM hive.test_schema.test_tbl;
   a    |            b
--------+--------------------------
 a_val2 | {ts=b2ts, topic=null}  <-- This is the important row, with the topic set to null
 a_val1 | {ts=b1ts, topic=b1topic}
(2 rows)

Query 20180625_221954_00001_94ydd, FINISHED, 1 node
Splits: 18 total, 18 done (100.00%)
0:03 [2 rows, 795B] [0 rows/s, 292B/s]
```
https://github.com/prestodb/presto/issues/10941

### CODE REVIEW

**Comment by electrum** [comment]
@dain This looks good to me. Can you think of any problems with this?

**Comment by dain** [comment]
Generally I think this is ok, but we likely need to add the other expected partition to file level coercions.  We can do this later, so my only remaining concern is how “spread out” this specific change is.

**Comment by shreyasjoshis** [comment]
> This doesn’t look right to me. It is only appending a single null to the block. I would expect one null per position.

Great catch @dain! Made this change. Also added tests.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java [hunk 1]
```diff
@@ -96,18 +97,28 @@ public Block readBlock(Type type)
...
-                StreamReader structField = structFields[i];
-                structField.prepareNextRead(nextBatchSize);
-                blocks[i] = structField.readBlock(typeParameters.get(i));
+                if (i < structFields.length) {
+                    StreamReader structField = structFields[i];
+                    structField.prepareNextRead(nextBatchSize);
+                    blocks[i] = structField.readBlock(typeParameters.get(i));
+                }
+                else {
+                    blocks[i] = getNullBlock(typeParameters.get(i), blocks[0].getPositionCount());
```
Comment (shreyasjoshis) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java]: This assumes `structFields.length >= 1`. Is there a way to avoid  this assumption?

**Comment by shreyasjoshis** [comment]
@dain Does this look good now?

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java [hunk 2]
```diff
@@ -177,4 +188,13 @@ public String toString()
...
                 .toString();
     }
+
+    private Block getNullBlock(Type type)
+    {
+        BlockBuilder blockBuilder = type.createBlockBuilder(null, 0);
+        for (int j = 0; j < nextBatchSize; j++) {
+            blockBuilder.appendNull();
+        }
+        return blockBuilder.build();
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java]:
```java
Block nullValueBlock = type.createBlockBuilder(null, 1)
        .appendNull()
        .build();
return new RunLengthEncodedBlock(nullValueBlock, positionCount)

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java [hunk 3]
```diff
@@ -322,32 +325,47 @@ public void testRoundTrip(Type type, List<?> readValues)
         }
     }
 
-    private void testStructRoundTrip(Type type, List<?> readValues)
+    private void testStructRoundTrip(Type type, List<?> values)
```
Comment (dain) [inline presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: Why rename this variable? If it does need to be renamed (I don't think it should), then it should be done in a separate commit to keep the review simple.

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java [hunk 4]
```diff
@@ -96,18 +97,28 @@ public Block readBlock(Type type)
...
-                    StreamReader structField = structFields[i];
-                    structField.prepareNextRead(nextBatchSize - nullValues);
-                    blocks[i] = structField.readBlock(typeParameters.get(i));
+                    if (i < structFields.length) {
+                        StreamReader structField = structFields[i];
+                        structField.prepareNextRead(nextBatchSize - nullValues);
+                        blocks[i] = structField.readBlock(typeParameters.get(i));
+                    }
+                    else {
+                        blocks[i] = getNullBlock(typeParameters.get(i));
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java]: This will create a block with one null entry per struct, but some of the structs are null.  Instead it should create a block with `nextBatchSize - nullValues` entries.   We can catch this kind of mistake with this after the `if/else`:
```java
        verify(Arrays.stream(blocks)
                .mapToInt(Block::getPositionCount)
                .distinct()
                .count() == 1);
```

File: presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java [hunk 5]
```diff
@@ -177,4 +195,12 @@ public String toString()
                 .addValue(streamDescriptor)
                 .toString();
     }
+
+    private Block getNullBlock(Type type, int positionCount)
```
Comment (dain) [inline presto-orc/src/main/java/com/facebook/presto/orc/reader/StructStreamReader.java]: I think this can be static

File: presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java (hunk 3 above)
Comment (dain) [reply presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java]: Ah I see.  That is ok then

**Comment by shreyasjoshis** [comment]
> One minor comment. Can you squash all of the commits into the first one, and I'll merge this (after the in-progress release goes out).

Done

**Comment by dain** [comment]
Oh can you update the commit title to "Allow missing fields in structs in ORC"

**Comment by dain** [comment]
@shreyasjoshis I mean the commit message title line... not the PR title :)

**Comment by shreyasjoshis** [comment]
Gentle bump in case this should have been merged

**Comment by dain** [comment]
Thanks, for the code (and reminder).
